lib/clusteraccess/advanced/clusteraccess.go

```diff
@@ -0,0 +1,1332 @@
+
+type clusterRegistrationBuilderImpl struct {
+	clusterRegistrationImpl
+}
+
+var _ ClusterRegistrationBuilder = &clusterRegistrationBuilderImpl{}
+
+// Build implements ClusterRegistrationBuilder.
+func (c *clusterRegistrationBuilderImpl) Build() ClusterRegistration {
+	return &c.clusterRegistrationImpl
+}
+
+// WithOIDCAccess implements ClusterRegistrationBuilder.
+func (c *clusterRegistrationBuilderImpl) WithOIDCAccess(cfg *clustersv1alpha1.OIDCConfig) ClusterRegistrationBuilder {
+	var f func(req reconcile.Request, additionalData ...any) (*clustersv1alpha1.OIDCConfig, error)
+	if cfg != nil {
+		f = func(req reconcile.Request, additionalData ...any) (*clustersv1alpha1.OIDCConfig, error) {
+			return cfg.DeepCopy(), nil
+		}
+	}
+	return c.WithOIDCAccessGenerator(f)
+}
+
+// WithOIDCAccessGenerator implements ClusterRegistrationBuilder.
+func (c *clusterRegistrationBuilderImpl) WithOIDCAccessGenerator(f func(req reconcile.Request, additionalData ...any) (*clustersv1alpha1.OIDCConfig, error)) ClusterRegistrationBuilder {
+	c.generateAccessOIDCConfig = f
+	if f != nil && c.generateAccessTokenConfig != nil {
+		// both access methods configured, disable token
+		c.generateAccessTokenConfig = nil
+	}
+	return c
+}
+
+// WithTokenAccess implements ClusterRegistrationBuilder.
+func (c *clusterRegistrationBuilderImpl) WithTokenAccess(cfg *clustersv1alpha1.TokenConfig) ClusterRegistrationBuilder {
+	var f func(req reconcile.Request, additionalData ...any) (*clustersv1alpha1.TokenConfig, error)
+	if cfg != nil {
+		f = func(req reconcile.Request, additionalData ...any) (*clustersv1alpha1.TokenConfig, error) {
+			return cfg.DeepCopy(), nil
+		}
+	}
+	return c.WithTokenAccessGenerator(f)
+}
+
+// WithTokenAccessGenerator implements ClusterRegistrationBuilder.
+func (c *clusterRegistrationBuilderImpl) WithTokenAccessGenerator(f func(req reconcile.Request, additionalData ...any) (*clustersv1alpha1.TokenConfig, error)) ClusterRegistrationBuilder {
+	c.generateAccessTokenConfig = f
+	if f != nil && c.generateAccessOIDCConfig != nil {
+		// both access methods configured, disable OIDC
+		c.generateAccessOIDCConfig = nil
+	}
+	return c
+}
+
+// WithScheme implements ClusterRegistrationBuilder.
+func (c *clusterRegistrationBuilderImpl) WithScheme(scheme *runtime.Scheme) ClusterRegistrationBuilder {
+	c.scheme = scheme
+	return c
+}
+
+// WithNamespaceGenerator implements ClusterRegistrationBuilder.
+func (c *clusterRegistrationBuilderImpl) WithNamespaceGenerator(f func(req reconcile.Request, additionalData ...any) (string, error)) ClusterRegistrationBuilder {
+	if f == nil {
+		f = DefaultNamespaceGenerator
+	}
+	c.generateNamespace = f
+	return c
+}
+
+// NewClusterRequest instructs the Reconciler to create and manage a new ClusterRequest.
+func NewClusterRequest(id, suffix string, generateClusterRequestSpec func(reconcile.Request, ...any) (*clustersv1alpha1.ClusterRequestSpec, error)) ClusterRegistrationBuilder {
```

Comment:
Please make the generateClusterRequestSpec function its own type and document what the parameters are. Especially the variable length of any parameters. What are they doing?

Reply:
Will do.